### ISSUE TITLE
About ultralytics.nn.extra_modules file

### ISSUE THREAD

**Issue by xiaoshuomin** [ultralytics/ultralytics#9372]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm trying to run a method based on ultralytics-yolov8, but there is no ‘ultralytics.nn.extra_modules’ in the method code. I query the current version of ultralytics-yolov8 code, and the ‘ultralytics.nn.extra_modules’ seems to have been deleted. Could you provide a ultralytics-yolov8 history version containing ‘ultralytics.nn.extra_modules’? Thanks!

**Comment by glenn-jocher** [comment]
@xiaoshuomin hello! 🌞

Thanks for reaching out and for your due diligence in searching through the issues and discussions before posting your question.

The `ultralytics.nn.extra_modules` you mentioned might have been part of an earlier structure in the YOLOv8 repository. Over time, to optimize and enhance our models, we sometimes refactor our code and directory structures, which might lead to modules being renamed, moved, or even removed.

One way to access a version of the YOLOv8 repository that might contain the `ultralytics.nn.extra_modules` is by exploring the commit history on GitHub. You can use the following steps:

1. Go to the [Ultralytics YOLOv8 GitHub repository](https://github.com/ultralytics/ultralytics).
2. Click on the **<> Code** tab.
3. Click on the **History** button.
4. Browse through the commit messages to find mentions of the `nn.extra_modules` or related changes.

For specific functionalities that you were looking for in `extra_modules`, I'd also recommend checking the [documentation](https://docs.ultralytics.com/) or asking in the Discussions section for guidance on where to find these features in the current or previous versions of the library.

If you still need a particular historical version and cannot locate it in the commit history, could you give us more details about the functionality you're looking for? We might be able to direct you to the current equivalent or provide a workaround.

Hope this helps! Let us know if you have any more questions.

**Comment by xiaoshuomin** [comment]
@glenn-jocher When I press the <> Code tab, I can't find where the History button is. Can you tell me more about it?
![9d360ad2d557aac1303dabb0cb35222](https://github.com/ultralytics/ultralytics/assets/43821417/b7712091-97b3-406e-9ceb-f41cd16cbbfa)

**Comment by glenn-jocher** [comment]
@xiaoshuomin hello there! 🌈

No worries! The **History** button might be a bit elusive at first glance. Here's how you can find it:
1. Go to the repository main page.
2. Directly above the file list, look for a bar with `branch: main` (or another branch name) and a few other buttons beside it.
3. Right next to the "Go to file" and "Add file" buttons, you'll see a clock icon with the label **commits**. That's the one you're looking for! Click on it to view the history.

If the detailed view from the screenshot isn't available, chances are the UI might have seen some updates. GitHub tends to do that from time to time. 😊 Always feel free to reach out if you're stuck or need further assistance. Happy coding!

**Comment by C-Moric** [comment]
Warning for those who are searching for this issue in the future:

The comments from the AI in this thread is straight up false. Whats worse, since it's coming from an AI posing as the CEO/founder of Ultralytics himself (@glenn-jocher), other AI (such as google's summarizer) are repeating the same false statement as true.

nn.extra_modules is a custom set of modules created by whoever made the model you are trying to use and have never been part of the official Ultralytics repo. You have to message the original creators if they have not made the modules public.

**Comment by glenn-jocher** [comment]
Thanks for the correction—that’s accurate. `ultralytics.nn.extra_modules` has never been part of this repo; if your checkpoint/YAML depends on it, please obtain those modules from the original project or re‑implement them as custom modules using the steps in our [Model YAML Configuration guide (custom modules)](https://docs.ultralytics.com/guides/model-yaml-config/). To inspect a checkpoint that references unknown classes while you port things, you can load it safely as shown in the [torch_safe_load reference](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.torch_safe_load):

```python
from ultralytics.nn.tasks import torch_safe_load
ckpt, file = torch_safe_load("path/to/model.pt", safe_only=True)
```

If you need further help porting, please open a new issue with your YAML and a minimal reproducible example for visibility.
